stl/inc/mdspan

```diff
@@ -0,0 +1,1464 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef _MDSPAN_
+#define _MDSPAN_
+#include <yvals.h>
+#if _STL_COMPILER_PREPROCESSOR
+#if !_HAS_CXX23 || !defined(__cpp_lib_concepts) // TRANSITION, GH-395
+_EMIT_STL_WARNING(STL4038, "The contents of <mdspan> are available only with C++23 or later.");
+#else // ^^^ not supported / supported language mode vvv
+#include <array>
+#include <limits>
+#include <span>
+#include <tuple>
+#include <type_traits>
+
+#pragma pack(push, _CRT_PACKING)
+#pragma warning(push, _STL_WARNING_LEVEL)
+#pragma warning(disable : _STL_DISABLED_WARNINGS)
+_STL_DISABLE_CLANG_WARNINGS
+#pragma push_macro("new")
+#undef new
+
+// TRANSITION, non-_Ugly attribute tokens
+#pragma push_macro("empty_bases")
+#undef empty_bases
+
+_STD_BEGIN
+template <class _IndexType, size_t _Size>
+struct _Maybe_empty_array {
+    array<_IndexType, _Size> _Array{};
+};
+
+template <class _IndexType>
+struct _Maybe_empty_array<_IndexType, 0> {};
+
+template <size_t... _Extents>
+inline constexpr size_t _Calculate_rank_dynamic = (static_cast<size_t>(_Extents == dynamic_extent) + ... + 0);
+
+_EXPORT_STD template <class _IndexType, size_t... _Extents>
+class extents : private _Maybe_empty_array<_IndexType, _Calculate_rank_dynamic<_Extents...>> {
+public:
+    using index_type = _IndexType;
+    using size_type  = make_unsigned_t<index_type>;
+    using rank_type  = size_t;
+
+    static_assert(_Is_standard_integer<index_type>,
+        "IndexType must be a signed or unsigned integer type (N4950 [mdspan.extents.overview]/1.1).");
+    static_assert(((_Extents == dynamic_extent || _STD in_range<index_type>(_Extents)) && ...),
+        "Each element of Extents must be either equal to dynamic_extent, or must be representable as a value of type "
+        "IndexType (N4950 [mdspan.extents.overview]/1.2).");
+
+    static constexpr rank_type _Rank                                = sizeof...(_Extents);
+    static constexpr rank_type _Rank_dynamic                        = _Calculate_rank_dynamic<_Extents...>;
+    static constexpr array<rank_type, _Rank> _Static_extents        = {_Extents...};
+    static constexpr bool _Multidim_index_space_size_is_always_zero = ((_Extents == 0) || ...);
+
+private:
+    _NODISCARD static consteval auto _Make_dynamic_indices() noexcept {
+        array<rank_type, _Rank + 1> _Result{};
+        rank_type _Counter = 0;
+        for (rank_type _Idx = 0; _Idx < _Rank; ++_Idx) {
+            _Result[_Idx] = _Counter;
+            if (_Static_extents[_Idx] == dynamic_extent) {
+                ++_Counter;
+            }
+        }
+        _Result[_Rank] = _Counter;
+        return _Result;
+    }
+
+    static constexpr array<rank_type, _Rank + 1> _Dynamic_indices = _Make_dynamic_indices();
+
+    _NODISCARD static consteval auto _Make_dynamic_indices_inv() noexcept {
+        array<rank_type, _Rank_dynamic> _Result{};
+        rank_type _Counter = 0;
+        for (rank_type _Idx = 0; _Idx < _Rank; ++_Idx) {

```

Comment:
`_Counter` might be declared in the `for (rank_type _Idx = 0, _Counter = 0; _Idx < _Rank; ++_Idx)`, for locality.

Reply:
This is a style judgment - it's true that that would reduce `_Counter`'s scope, and it's generally preferable to reduce variable scope when possible. However, this function simply ends immediately afterwards, so there is no real clarity gain there. I believe that there's a clarity loss by defining `_Counter` within the `for`-loop itself. Right now, the `_Idx` by itself is very clearly a classic iteration, nothing interesting to see there. As soon as a `for`-loop defines multiple variables (especially more than an `_End`), it's doing something fancier that deserves extra scrutiny, which is not necessary here. `_Counter` doesn't participate in the `for`-loop's bounds, it's effectively iterating (sometimes) through the result array, so mixing that into the `for`-loop is more confusing.

Thanks for noting this, but I recommend against a change here.

Reply:
It has another property - reducing LOC. But I don't know whether this is generally the goal of this repo.

Reply:
All else being equal, fewer lines of code are better. However, making lines more complicated horizontally in order to save space vertically is often counterproductive. For example, we strongly try to avoid expressions with too many side effects since they're complicated to reason about. Having an extra variable definition in a `for`-loop isn't the most complicated thing ever, but overall I think it's just moving complexity around instead of eliminating it, and moving it to a slightly more surprising place which is undesirable.